DPL: wait as long as possible for Sporadic inputs

ktf opened the pull request:
DPL: wait as long as possible for Sporadic inputs

Right now if we have the standard consumeWhenAll policy and we have a sporadic
input, it will wait indefinitely until all the inputs arrive or it will drop
timeframes without that sporadic input.

This changes the behavior and waits only until the oldest possible timeframe
does not allow the Sporadic input to be there. At which point, it schedules the
processing in any case, under the assumption that a task declaring a sporadic
input knows what to do in case it's not there.

ktf commented:
@shahor02 @davidrohr as we discussed at some point. I have not tested it yet, though.

alibuild commented:
Error while checking build/O2/fullCI for bdecacebf99acbac801cdbc35e45973a9f32b600 at 2023-12-13 06:05:
```
## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/QualityControl/v1.127.1/v1.127.1/Framework/src/TaskRunnerFactory.cxx:203:107: error: no matching function for call to 'o2::framework::CompletionPolicy::CompletionPolicy(<brace-enclosed initializer list>)'
ninja: build stopped: cannot make progress due to previous errors.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12451/bdecacebf99acbac801cdbc35e45973a9f32b600/build_O2_fullCI/pretty.html).

davidrohr commented on Framework/Core/src/CompletionPolicyHelpers.cxx:
```diff
@@ -107,10 +108,32 @@ CompletionPolicy CompletionPolicyHelpers::consumeWhenAll(const char* name, Compl
         return CompletionPolicy::CompletionOp::Wait;
       }
+      if (input.header == nullptr && spec.lifetime == Lifetime::Sporadic) {
+        missingSporadic |= true;
```
why |=, just use =

ktf replied:
```suggestion
        missingSporadic = true;
```

davidrohr reviewed (dismissed):
looks good, just have a minor comment

alibuild commented:
Error while checking build/O2/fullCI for 8642aeeac6d26535a84e2dd101a0bd148e5f417c at 2023-12-13 23:15:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
command o2-sim-digitizer-workflow -n 5 --simPrefixQED qed/o2sim --qed-x-section-ratio 4417.27 --disable-mc --sims o2sim --tpc-lanes 32 --shm-segment-size 8000000000 -b --skipDet TRD --configKeyValues "HBFUtils.nHBFPerTF=128;HBFUtils.orbitFirst=0;HBFUtils.orbitFirstSampled=0;HBFUtils.obligatorySOR=false;HBFUtils.runNumber=310000;;HBFUtils.maxNOrbits=128;" --interactionRate 50000 had nonzero exit code 128
[ERROR] Workflow crashed - PID 29308 (EMCALDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29339 (ZDCDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29869 (GRPUpdater) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29381 (PHOSDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29325 (FT0Digitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29316 (FDDDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29347 (CPVDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29372 (MIDDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29353 (ITSDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR] Workflow crashed - PID 29363 (MFTDigitizer) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR]  - Device EMCALDigitizer: pid 29308 (exit 128)
[ERROR]  - Device FDDDigitizer: pid 29316 (exit 128)
[ERROR]  - Device FT0Digitizer: pid 29325 (exit 128)
[ERROR]  - Device ZDCDigitizer: pid 29339 (exit 128)
[ERROR]  - Device CPVDigitizer: pid 29347 (exit 128)
[ERROR]  - Device ITSDigitizer: pid 29353 (exit 128)
[ERROR]  - Device MFTDigitizer: pid 29363 (exit 128)
[ERROR]  - Device MIDDigitizer: pid 29372 (exit 128)
[ERROR]  - Device PHOSDigitizer: pid 29381 (exit 128)
[ERROR]  - Device GRPUpdater: pid 29869 (exit 128)
[ERROR] SEVERE: Device EMCALDigitizer (29308) returned with 128


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12451/8642aeeac6d26535a84e2dd101a0bd148e5f417c/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for a59ea0ead75c70e355ac5b4e7eea23a86b4dcdd8 at 2024-01-12 13:52:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
command /sw/slc8_x86-64/O2/12451-slc8_x86-64-local2/prodtests/full-system-test/dpl-workflow.sh had nonzero exit code 128
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data HMP/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/0 was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data MID/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data CTP/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data TOF/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data TPC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data ITS/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data MFT/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data MCH/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data PHS/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data TRD/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data CPV/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data EMC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FDD/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FT0/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FV0/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data ZDC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/52443 was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data HMP/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/0 was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data MID/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data CTP/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data TOF/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data TPC/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data ITS/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48415:raw-file-reader]: [12:51:27][ERROR] Expected Lifetime::Timeframe data MFT/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12451/a59ea0ead75c70e355ac5b4e7eea23a86b4dcdd8/build_O2_fullCI/pretty.html).

ktf committed:
DPL: wait as long as possible for Sporadic inputs

Right now if we have the standard consumeWhenAll policy and we have a sporadic
input, it will wait indefinitely until all the inputs arrive or it will drop
timeframes without that sporadic input.

This changes the behavior and waits only until the oldest possible timeframe
does not allow the Sporadic input to be there. At which point, it schedules the
processing in any case, under the assumption that a task declaring a sporadic
input knows what to do in case it's not there.

alibuild commented:
Error while checking build/O2/fullCI for 799976301a2984351675b5da7973fa9f2dcc4423 at 2024-01-14 14:07:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
command /sw/slc8_x86-64/O2/12451-slc8_x86-64-local5/prodtests/full-system-test/dpl-workflow.sh had nonzero exit code 128
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data HMP/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/0 was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MID/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data CTP/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TOF/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TPC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data ITS/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MFT/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MCH/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data PHS/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TRD/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data CPV/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data EMC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FDD/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FT0/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FV0/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data ZDC/RAWDATA was not created for timeslice 0 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/52443 was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data HMP/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FLP/DISTSUBTIMEFRAME/0 was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MID/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data CTP/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TOF/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TPC/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data ITS/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MFT/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data MCH/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data PHS/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data TRD/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data CPV/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data EMC/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FDD/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FT0/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data FV0/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48004:raw-file-reader]: [13:06:46][ERROR] Expected Lifetime::Timeframe data ZDC/RAWDATA was not created for timeslice 1 and might result in dropped timeframes
[48031:BadMapCalibSpec]: [13:07:00][ERROR] Insufficient statistics: 3 entries in lowE histo, do nothing
[48089:ctf-writer]: [13:07:02][ERROR] Some Lifetime::Timeframe data got dropped starting at 1
[ERROR] Workflow crashed - PID 48079 (tpc-calibrator-gainmap-tracks) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR]  - Device tpc-calibrator-gainmap-tracks: pid 48079 (exit 128)
[ERROR] SEVERE: Device tpc-calibrator-gainmap-tracks (48079) returned with 128
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12451/799976301a2984351675b5da7973fa9f2dcc4423/build_O2_fullCI/pretty.html).

ktf commented:
Checked on hyperloop as well.

martenole commented:
Do we understand why this PR passed the fullCI although (at least for me) it seems to always crash in the `tpc-calibrator-gainmap-tracks`?

Edit: it has been fixed in the meantime, I just wanted to understand why the CI didn't catch it. Or could it have been by chance due to the timing of the messages that it worked?

davidrohr commented:
For me the crash in the calibrator-gainmap-tracks happens only randomly. In case of the CI run for this PR, it was green.

davidrohr commented:
@martenole  : This has fixed some of the issues: https://github.com/AliceO2Group/AliceO2/pull/12536
If you have a reproducer (since you are currently using epn000), could you check if that fixes it?
